### PR TITLE
Fix `IO::FileDescriptor.new` for closed fd

### LINKED ISSUE #14569: Crystal runtime crashes on closed file descriptor for STDIN
As part of programm startup, the Crystal runtime initializes handles for the three standard streams.
If such a stream is already closed, the runtime crashes.

This affects any Crystal program which uses `STDIN` *somehow* (i.e. the constant is initialized).

A minimal reproduction is `crystal <&-` (the shell redirection `<&-` instructs that stdin is to be closed). This results in the following error:
```
Unhandled exception: Unable to get info: Bad file descriptor (IO::Error)
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in 'raise<IO::Error>:NoReturn'
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in 'Crystal::System::FileDescriptor::system_info<Int32>:File::Info'
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in 'IO::FileDescriptor::new<Int32>:IO::FileDescriptor'
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in 'IO::FileDescriptor::from_stdio<Int32>:IO::FileDescriptor'
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in '__crystal_main'
  from /opt/homebrew/Cellar/crystal/1.12.1_1/bin/crystal in 'main'
```

The same error does *not* appear when passing a closed output file descriptor (`crystal >&-`).

I'm not sure how to reproduce in pure Crystal yet. Passing a closed `IO` to `Process.run` errors on the invoking process (`Closed stream` from `Process#copy_io`). This needs to be fixed as well.
```cr
stdin = IO::Memory.new
stdin.close
Process.run("crystal", input: stdin, output: :inherit, error: :inherit)
```

This was originally reported in https://forum.crystal-lang.org/t/unable-to-get-info-bad-file-descriptor-io-error/6820

### PR DESCRIPTION
**Pull request by straight-shoota**
Skips any operations for detecting/setting blocking mode in `IO::FileDescriptor.new` when the file descriptor is closed. In this case it doesn't matter anyway.

Resolves #14569

### CODE REVIEW

File: src/io/file_descriptor.cr [hunk 1]
```diff
@@ -43,17 +43,19 @@ class IO::FileDescriptor < IO
...
 
-    if blocking.nil?
-      blocking =
-        case system_info.type
-        when .pipe?, .socket?, .character_device?
-          false
-        else
-          true
-        end
+    unless @closed
```
Comment (ysbaddaden) [inline src/io/file_descriptor.cr:46]: **nitpick**: maybe `return if @closed`

**Review by ysbaddaden — APPROVED** [review]

**Comment by ysbaddaden** [comment]
Oh, the spec fails on Windows:

> Unable to get info (Crystal::System::FileDescriptor): The handle is invalid. (IO::Error)

**Comment by straight-shoota** [comment]
The issue was that `system_closed?` was not implemented on Windows. I fixed that.
Now this is not ideal because with `blocking.nil?` and an open handle, `GetFileType` is called twice (once in `system_closed?` and once in `system_info`). Perhaps we could refactor `FileDescriptor#initialize` to avoid that (i.e. delegate a larger chung to the system implementations).

**Review by ysbaddaden — APPROVED** [review]

**Comment by ysbaddaden** [comment]
AFAIK we don't call `IO::FileDescriptor.new` much with `blocking` being nil in practice.  I think only `.from_stdio` does.

**Comment by straight-shoota** [comment]
`blocking = nil` is the default value for `IO::FileDescriptor.new`. But of course, this constructor is not used much. It's also possible to pass `blocking: nil` to `File.new` from user code, though.
But it's probably not used much and the gain is minimal. So no rush to optimize this. But it could be an enhancement somtime in the future. `system_closed?` is only called from the constructor, so it could easily lend itself for a refactor.

**Pull request merged**